src/analyses/threadFlag.ml

```diff
@@ -0,0 +1,87 @@
+(** Multi-threadedness analysis. *)
+
+module GU = Goblintutil
+module LF = LibraryFunctions
+
+open Prelude.Ana
+open Analyses
+
+let is_multi_ask (ask: Queries.ask): bool =
+  match ask Queries.SingleThreaded with
+  | `Bool b -> not b
+  | `Top -> true
+  | _ -> failwith "is_multi_ask"
+let is_multi ctx: bool = is_multi_ask ctx.ask
+
+
+module Spec =
+struct
+  include Analyses.DefaultSpec
+
+  module Flag = ConcDomain.Simple
+  module D = Flag
+  module C = Flag
+  module G = Lattice.Unit
+
+  let name () = "threadflag"
+
+  let startstate v = Flag.bot ()
+  let exitstate  v = Flag.get_multi ()
+
+  let morphstate v _ = Flag.get_single ()
+
+  let create_tid v =
+    Flag.get_multi ()
+
+  let body ctx f = ctx.local
+
+  let branch ctx exp tv = ctx.local
+
+  let return ctx exp fundec  =
+    match fundec.svar.vname with
+    | "__goblint_dummy_init" ->
+      (* TODO: is this necessary? *)
+      Flag.join ctx.local (Flag.get_main ())
+    | "StartupHook" ->
+      (* TODO: is this necessary? *)
+      Flag.get_multi ()
+    | _ ->
+      ctx.local
+
+  let assign ctx (lval:lval) (rval:exp) : D.t  =
+    ctx.local
+
+  let enter ctx lval f args =
+    [ctx.local,ctx.local]
+
+  let combine ctx lval fexp f args fc st2 = st2
+
+  let special ctx lval f args =
+    ctx.local
+
+  let query ctx x =
+    match x with
+    | Queries.SingleThreaded -> `Bool (Queries.BD.of_bool (not (Flag.is_multi ctx.local)))
+    | Queries.IsNotUnique -> `Bool (Flag.is_bad ctx.local)
+    (* This used to be in base but also commented out. *)
+    (* | Queries.IsPublic _ -> `Bool (Flag.is_multi ctx.local) *)
```

Comment:
Uncommenting this would affect the privatization of `precious_globals` (which is in turn somehow related to `early_globs`). Also this should probably only return `Bool false` if `Flag.is_multi ctx.local` is `false` and `Top` otherwise. 

https://github.com/goblint/analyzer/blob/813b9ee72a73bdf486356063cf6f8c9546bf8cf3/src/analyses/base.ml#L36-L40

Reply:
I think it should be fine as is, because in the boolean lattice used for these results `Bool true` is the top of that type. I've had to carefully consider the behavior of the related boolean queries because of this to make sure that now with multiple analyses answering the thread-related ones, they meet in the right way.

As for this commented out query, I guess the idea at some point was that in single threaded mode everything should be private anyway. But that is checked in base's `is_private` now, so the query wasn't needed. I'll leave it commented out here though because I suppose it could be changed back (along with the precious globs check).